File: python/taflow/metrics/system_quality_number.py
"""Closed-trade System Quality Number metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import SystemQualityNumber as _Native
from ._input import as_metric_series


class SystemQualityNumber:
    """Compute System Quality Number from realized closed-trade P&L.

    The formula is ``sqrt(n) * mean(trade_pnl) / sample_std(trade_pnl)``
    with ``ddof=1``. Warm-up requires two usable trades, and a zero sample
    standard deviation makes the result undefined, so ``value`` and
    ``compute`` return ``None`` in either case. A defined result retains the
    sign of mean trade P&L.

    The primary oracle mapping is vectorbt 0.28.1 ``Trades.sqn``. Its source
    tarball is pinned by SHA-256
    ``aceeb4767a1bd5be18329bc85779b2fc744b1edc4513ad19c4dbc3fc7d83d301``;
    vectorbt is not installed in the test environment, so NumPy 2.4.6
    ``mean`` and ``std(ddof=1)`` provide the executable independent numerical
    cross-check without constructing synthetic portfolio records.

    Only ``from_trades`` is available, and each input is realized P&L for one
    chronological closed trade. Period returns, equity, and period P&L are not
    accepted or annualized. Rust owns numerically stable Welford moments and
    O(1)-memory state. NaNs are omitted by default or rejected with
    ``nan_policy="raise"``; infinities are always rejected. Mutating lifecycle
    methods are fluent and bulk execution releases the GIL.
    """

    def __init__(self) -> None:
        """Reject ambiguous construction; use ``from_trades``."""
        raise TypeError("use SystemQualityNumber.from_trades")

    @classmethod
    def from_trades(
        cls,
        trades: Any,
        *,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "SystemQualityNumber":
        """Construct from realized P&L for chronological closed trades."""
        state = cls.__new__(cls)
        state._state = _Native("trades", nan_policy)
        return state.extend(trades, column=column)

    def append(self, trade_pnl: float) -> "SystemQualityNumber":
        """Append one realized closed-trade P&L and return this metric."""
        self._state.append(float(trade_pnl))
        return self

    def extend(
        self, trades: Any, *, column: str | None = None
    ) -> "SystemQualityNumber":
        """Append chronological closed-trade P&L and return this metric."""
        self._state.extend(as_metric_series(trades, column=column))
        return self

    @property
    def value(self) -> float | None:
        """Return SQN, or ``None`` for insufficient or constant trade P&L."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current scalar without replaying processed trades."""
        return self._state.compute()

    def reset(self) -> "SystemQualityNumber":
        """Clear trades, preserve configuration, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable closed trades processed by Rust."""
        return len(self._state)


__all__ = ["SystemQualityNumber"]

File: crates/taflow-metrics/src/metrics/system_quality_number.rs
use crate::{
    primitives::OnlineMoments, MetricError, MetricInputKind, MetricInputState, MetricResult,
    NanPolicy,
};

/// Square root of trade count times mean trade P&L divided by sample deviation.
#[derive(Debug, Clone)]
pub struct SystemQualityNumber {
    input: MetricInputState,
    moments: OnlineMoments,
    trade_sum: f64,
    trade_sum_compensation: f64,
}

impl SystemQualityNumber {
    /// Construct an empty state for realized closed-trade P&L.
    pub fn new(nan_policy: NanPolicy) -> MetricResult<Self> {
        Ok(Self {
            input: MetricInputState::unbound(nan_policy),
            moments: OnlineMoments::new(),
            trade_sum: 0.0,
            trade_sum_compensation: 0.0,
        })
    }

    /// Append one chronological closed-trade P&L and return the current SQN.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(trade_pnl) = self.input.append(value)? {
            self.moments.append(trade_pnl);
            let updated_sum = self.trade_sum + trade_pnl;
            if self.trade_sum.abs() >= trade_pnl.abs() {
                self.trade_sum_compensation += (self.trade_sum - updated_sum) + trade_pnl;
            } else {
                self.trade_sum_compensation += (trade_pnl - updated_sum) + self.trade_sum;
            }
            self.trade_sum = updated_sum;
        }
        Ok(self.value())
    }

    /// Append a chronological slice through the same persistent state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        for &value in values {
            self.append(value)?;
        }
        Ok(self.value())
    }

    /// Return SQN, or `None` with fewer than two trades or zero sample deviation.
    pub fn value(&self) -> Option<f64> {
        let standard_deviation = self.moments.standard_deviation(1)?;
        if standard_deviation == 0.0 || !standard_deviation.is_finite() {
            return None;
        }
        let mean = (self.trade_sum + self.trade_sum_compensation) / self.moments.len() as f64;
        let result = (self.moments.len() as f64).sqrt() * mean / standard_deviation;
        result.is_finite().then_some(result)
    }

    /// Return the current result without replaying prior trades.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while preserving the closed-trade domain.
    pub fn reset(&mut self) {
        self.input.reset();
        self.moments.reset();
        self.trade_sum = 0.0;
        self.trade_sum_compensation = 0.0;
    }

    /// Return the number of usable closed trades processed.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable closed trades have been processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}

crate::impl_trades_only_metric_lifecycle!(SystemQualityNumber);
